db/migrate/20190320145911_change_solar_park_inputs_to_capacity.rb

```diff
@@ -0,0 +1,67 @@
+class ChangeSolarParkInputsToCapacity < ActiveRecord::Migration[5.1]
+  INPUTS = {
+    energy_power_solar_pv_solar_radiation: 20.0
+  }.freeze
+
+  def change
+    reversible do |dir|
+      dir.up do
+        update_scenarios(INPUTS, 'number_of', 'capacity_of')
+      end
+
+      dir.down do
+        update_scenarios(
+          INPUTS.transform_values { |v| 1 / v },
+          'capacity_of',
+          'number_of'
+        )
+      end
+    end
+  end
+
+  private
+
+  def update_scenarios(collection, from_prefix, to_prefix)
+    total = scenarios.count
+    changed = 0
+
+    say "Checking and migrating #{total} scenarios"
+
+    scenarios.find_each.with_index do |scenario, index|
+      updated = collection.reduce(false) do |memo, (key, multiplier)|
+        update_input(
+          scenario,
+          "#{from_prefix}_#{key}",
+          "#{to_prefix}_#{key}",
+          multiplier
+        ) || memo
+      end
+
+      if updated
+        scenario.save(validate: false)
+        changed += 1
+      end
+
+      say "#{index + 1}/#{total} (#{changed} migrated)" if (index % 1000).zero?
+    end
+
+    say "#{total}/#{total} (#{changed} migrated)"
+  end
+
+  # Changes an input in a `scenario` from the `from` key to the `to` key,
+  # multiplying the value by `multiplier`.
+  #
+  # If the `from` key is not present, returns a falsey value.
+  def update_input(scenario, from, to, multiplier)
+    if scenario.user_values.key?(from)
+      scenario.user_values[to] = scenario.user_values.delete(from) * multiplier
+      true
+    end
+  end
+
+  # All protected scenarios, and any unprotected scenarios since Jan 1st 2018
+  # will be updated.
+  def scenarios
+    Scenario.migratable_since(Date.new(2018, 1, 1))
```

Comment:
Out of interest, why are all scenarios from the beginning of 2018 being migrated? Was this an intentional choice, or just copy/paste from another migration?

Normally we just do protected scenarios, and unprotected scenarios from the past month. In that case, you can just use `Scenario.migratable`:

```diff
  def scenarios
-   Scenario.migratable_since(Date.new(2018, 1, 1))
+   Scenario.migratable
  end
```

Reply:
It's not really an intentional choice, actually. I've indeed copy/pasted the method from a similar migration for the hydrogen sliders that was performed recently. Anyways, thanks for the explanation. I'll adopt the suggested changes!

Reply:
> Normally we just do protected scenarios, and unprotected scenarios from the past month.

I think it's safer to take scenarios from the past 3 months or so if that's possible performance wise.

@antw Do you know if the unsaved scenarios that you migrated manually a couple of weeks ago (e.g. https://github.com/quintel/etengine/blob/master/db/migrate/20190226130006_update_housing_stock_and_insulation_inputs_om.rb) have been marked as protected now? If not these migrations won't apply to them

Reply:
> @antw Do you know if the unsaved scenarios that you migrated manually a couple of weeks ago have been marked as protected now?

They are indeed now protected.